### PR TITLE
migrate to preact-x

### PR DESCRIPTION
**Pull request by ForsakenHarmony**

### CODE REVIEW

File: src/Components/RequestGeneratorBuilder.js
Comment (ForsakenHarmony) [inline src/Components/RequestGeneratorBuilder.js:188]: I think the more idiomatic way to do this would be something like a render prop

https://reactjs.org/docs/render-props.html

(You pass in a function to generate the children and pass the components to that instead of receiving children and replacing placeholder)

**Comment by ForsakenHarmony** [comment]
One of the main changes is that `setState` is async now, I'm not sure I found all cases yet where this broke something

File: src/Components/FlashMessage.js
Comment (ForsakenHarmony) [inline src/Components/FlashMessage.js:48]: this seems to only work once, not sure why right now

**Comment by mal-tee** [comment]
Thank you very much!

Atm the cypress tests don't run because the linter fails beforehand. You could disable the linter (`react/prop-test`) to see if the tests run through. :D

File: src/Components/RequestGeneratorBuilder.js
Comment (ForsakenHarmony) [reply src/Components/RequestGeneratorBuilder.js]: I've implemented this now

**Comment by baltpeter** [comment]
Wow, thank you so much! That is immensely helpful.

I'll try my best to review this quickly but the review is probably going to take some time due to the sheer amount of files touched in the PR.

File: webpack.common.js
Comment (baltpeter) [inline webpack.common.js:24]: Not sure if this configuration is a good fit for us, to be honest. While I agree that duplicating code across the different entrypoints isn't great, we are trying pretty hard to reduce the code that needs to be downloaded per page. We have a few pages (mostly those with a generator embedded) that are *really* heavy but most of our pages are optimized to load as little JS as possible. Currently, Webpack 'only' complains about the asset sizes in the three generator-related files:

image

With this change, *every* entrypoint is now a problem:

image

I am especially worried about this because of our blog and company single pages. They are pretty optimized to load as quickly as possible and adding that much (unnecessary) JS to them
would potentially hurt our SEO, which we need to avoid.

File: webpack.common.js
Comment (baltpeter) [inline webpack.common.js:45]: This is the first time I am seeing this option, so I may well be wrong. But we have some code that depends on the 'raw' (untranspiled) JS added using Hugo and the Webpack-transpiled JS having access to the same `window` object. Is this option a good idea in this case?

Comment (baltpeter) [reply webpack.common.js:45]: Although, now that I think about it some more, in those instances we are already explicitly using `window.something` instead of just `something`, so this may not apply.

File: webpack.common.js
Comment (baltpeter) [inline webpack.common.js:52]: Is there any benefit to using this loader? We are currently thinking about switching our JS build from Webpack to Hugo with esbuild, so I would like to avoid adding Webpack-specific features that we later need to remove again/rebuild. :D

Comment (baltpeter) [reply webpack.common.js:52]: If we *do* keep that, the file names need to be adjusted. The currently used `pdf.worker.gen.worker.js` is not covered by our `.gitignore`.

File: package.json
Comment (baltpeter) [inline package.json:120]: I am assuming you have removed that because (on some systems) this takes quite a while? :D I would still prefer to keep it as we did accidentally commit merge conflict artifacts in the past. @mal-tee already complained about the speed and is working on an implementation that can leverage ripgrep.

**Review by baltpeter — COMMENTED** [review]
Huge thanks once again. Below are my comments. Let me know what you think.

**Comment by baltpeter** [comment]
@mal-tee I have also requested a review from you. Let us know if you find any regressions that we may have missed.

File: layouts/partials/scripts.html
Comment (mal-tee) [reply layouts/partials/scripts.html:32]: Not sure if this is the same, but if you click on a category,e.g. https://deploy-preview-485--datenanfragen.netlify.app/categories/nonprofit/ an error is thrown and the search bar isn't rendered.

```
**Debug information:**

message := TypeError: P is null
type := error
colno := 4457
filename := https://deploy-preview-485--datenanfragen.netlify.app/js/company-list.gen.js
lineno := 14
error.message := P is null
error.name := TypeError
error.stack := 117/<@https://deploy-preview-485--datenanfragen.netlify.app/js/company-list.gen.js:14:4457
117@https://deploy-preview-485--datenanfragen.netlify.app/js/company-list.gen.js:14:4651
s@https://deploy-preview-485--datenanfragen.netlify.app/js/company-list.gen.js:14:564
n@https://deploy-preview-485--datenanfragen.netlify.app/js/company-list.gen.js:14:433
@https://deploy-preview-485--datenanfragen.netlify.app/js/company-list.gen.js:14:1502
@https://deploy-preview-485--datenanfragen.netlify.app/js/company-list.gen.js:14:1506

error.context := defaultPrevented := false
eventPhase := 2
isTrusted := true
returnValue := true
code_version := 1.0.0
user_agent := Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0
url.href := https://deploy-preview-485--datenanfragen.netlify.app/categories/nonprofit/
url.origin := https://deploy-preview-485--datenanfragen.netlify.app
url.protocol := https:
url.host := deploy-preview-485--datenanfragen.netlify.app
url.hostname := deploy-preview-485--datenanfragen.netlify.app
url.port := url.pathname := /categories/nonprofit/
url.search := url.hash := 
```

**Comment by mal-tee** [comment]
### Switching the request type doesn't seem to work. 

To reproduce: Visit /generator, click "Erasure Request", "Send email", "Copy manually" --> The request body is for an access request instead of an erasure request.

File: layouts/partials/scripts.html
Comment (baltpeter) [reply layouts/partials/scripts.html:32]: @mal-tee That's because of:

https://github.com/datenanfragen/website/blob/0dfdd5275e22237c3dbd895a2a79ccdef32076f1/src/company-list.js#L139

It doesn't find the element and we of course cannot access `(null).parentElement`.

File: package.json
Comment (ForsakenHarmony) [reply package.json:120]: Didn't work on Windows at all, didn't mean to commit this though

File: webpack.common.js
Comment (ForsakenHarmony) [reply webpack.common.js:45]: This was something I ran into debugging the loader

File: webpack.common.js
Comment (ForsakenHarmony) [reply webpack.common.js:24]: Didn't check the build output, this was definitely not the intention

File: src/Components/ActWidget.js
Comment (ForsakenHarmony) [reply src/Components/ActWidget.js:7]: Yes, every entry seems to create a completely separate instance of `preact`

File: webpack.common.js
Comment (ForsakenHarmony) [reply webpack.common.js:52]: I suppose I can get rid of it, fixed an issue I was having with split chunks and makes the worker a bit nicer to work with (automatically adds the entry, nice loading by importing)

File: src/Components/DonationWidget.js
Comment (ForsakenHarmony) [reply src/Components/DonationWidget.js:1]: Yes, another problem with `setState` being async now

File: src/Components/ActWidget.js
Comment (ForsakenHarmony) [reply src/Components/ActWidget.js]: This is fixed now

File: src/Components/Generator/DynamicInput.js
Comment (ForsakenHarmony) [reply src/Components/Generator/DynamicInput.js:334]: Should be fixed

File: src/Components/DonationWidget.js
Comment (ForsakenHarmony) [reply src/Components/DonationWidget.js:1]: Should be fixed

**Comment by ForsakenHarmony** [comment]
@mal-tee that should be fixed as well now

**Review by baltpeter — CHANGES_REQUESTED** [review]
Alright, finally got around to looking into the new changes in detail. Huge thanks once again! These are some really good changes, even regardless of the move to Preact X. I definitely feel like the code is a lot more robust now.

I have left some minor comments below, other than that, this is pretty much ready for merge. Also:

* Feel free to add yourself to the [`AUTHORS` file](https://github.com/datenanfragen/website/blob/master/AUTHORS) if you want. That way, you are listed on our [open source page](https://www.datarequests.org/open-source#contributors) as a contributor.
* @mal-tee Since there have been some quite substantial changes since your last review and our tests are pretty far from complete at the moment, would you mind doing another thorough manual testing pass over everything?

**Comment by mal-tee** [comment]
Will do!

File: src/id-data-controls.js
Comment (mal-tee) [reply src/id-data-controls.js:167]: We could enumerate them: "Custom field 1", "Custom field 2" / "Eigenes Feld 1", "Eigenes Feld 2"..

File: src/Components/Generator/DynamicInputContainer.js
Comment (mal-tee) [reply src/Components/Generator/DynamicInputContainer.js:162]: This also occurs with every other kind of field. Can also reproduced by removing the default (prefilled) field and re-adding it.

**Review by mal-tee — DISMISSED** [review]
I didn't find any additional bugs. :D

Thanks again!

File: src/id-data-controls.js
Comment (ForsakenHarmony) [reply src/id-data-controls.js:210]: Yeah, not happy with it either, not sure why you have the complicated logic to store all these in indexeddb individually, could have 3 entries "fixed", "custom" and "signature" in there instead

Comment (baltpeter) [reply src/id-data-controls.js:210]: @zner0L Could you please take a look at this since you implemented this logic?

Comment (ForsakenHarmony) [reply src/id-data-controls.js:210]: :upside_down_face:

Comment (baltpeter) [reply src/id-data-controls.js:210]: Sorry about the delay. :/ I'll ask @zner0L again.

File: src/Components/CommentsWidget.js
Comment (zner0L) [inline src/Components/CommentsWidget.js:142]: Why allow strings here?

File: src/id-data-controls.js
Comment (zner0L) [inline src/id-data-controls.js:293]: This is actually a really nice way to solve this. However, I think we no longer need the events anyway, since we retired the id data widget we had on top of the page. But now that you implemented it…

File: src/id-data-controls.js
Comment (zner0L) [reply src/id-data-controls.js:210]: Ok, sorry it took me so long! I don't think there is a strong reason not to have these stored in separate tables (except maybe for needing more then one localforage instance). Or are you talking about saving the JSON to the database directly? That sounds very reasonable tbh! As far as I am concerned, this issue wasn't caused by your PR so we might as well merge and resolve this in another issue.

**Review by zner0L — CHANGES_REQUESTED** [review]
Ok, first of all: Thank you so much for doing this! This is a major improvement of our code and it'll be much more robust after this. I am sorry you had to see all our early code I guess :D Second of all: I am sorry it took me so long to fight through my mental health issues to review your PR, you would've deserved better!
Finally, I basically only have some minor questions and other preferences, but other than that I can only approve. As far as I am concerned, we can merge this if you are too annoyed to address my questions.

File: src/Components/RequestGeneratorBuilder.js
Comment (baltpeter) [reply src/Components/RequestGeneratorBuilder.js:321]: I don't think that's a problem. Imo it makes sense that a call to `changeRequest()` would trigger that behaviour. And yeah, it's good to leave less room for errors.

File: src/Components/CommentsWidget.js
Comment (baltpeter) [reply src/Components/CommentsWidget.js:142]: Currently, at least, that's what the backend returns: https://backend.datenanfragen.de/comments/get/en/company/epicgames

File: src/Components/StarWidget.js
Comment (baltpeter) [reply src/Components/StarWidget.js:52]: See above.

**Review by baltpeter — APPROVED** [review]
Huge thanks once again, @ForsakenHarmony.

I have now addressed the remaining review feedback. In the interest of getting this merged quickly, I have opened #553 for the weird behaviour with the custom inputs in the ID data controls.

File: src/Components/CommentsWidget.js
Comment (zner0L) [reply src/Components/CommentsWidget.js:142]: Interesting :D Well, then the error is on our side.

**Pull request merged**
